airflow/models.py

```diff
@@ -3536,11 +3536,14 @@ def set(
         session.expunge_all()
 
         # remove any duplicate XComs
-        session.query(cls).filter(
+        x = session.query(cls).filter(
             cls.key == key,
             cls.execution_date == execution_date,
             cls.task_id == task_id,
-            cls.dag_id == dag_id).delete()
+            cls.dag_id == dag_id).first()
+
+        if x:
+            session.delete(x)
```

Comment:
+1 for `update` vs `delete`.. it'll also saves the cost of db indexing.. [i.e.](http://stackoverflow.com/questions/824936/mysql-performance-delete-or-update)